test/suite/Annotations.test.php

```diff
@@ -585,12 +590,149 @@ protected function testGetAnnotationsFromTrait()
             return;
         }
 
-        eval('trait TestTrait { }');
-        $this->setExpectedException(
-            self::ANNOTATION_EXCEPTION,
-            "Reading annotations from interface/trait 'TestTrait' is not supported"
-        );
-        Annotations::ofClass('TestTrait', '@note');
+        $annotations = Annotations::ofClass('SimpleTrait', '@note');
+
+        $this->check(count($annotations) === 1, 'SimpleTrait has one note annotation.');
+    }
+
+    protected function testCanGetMethodAnnotationsIncludedFromTrait()
+    {
+        if (version_compare(PHP_VERSION, '5.4.0', '<')) {
+            $this->pass();
+            return;
+        }
+
+        $annotations = Annotations::ofMethod(new \ReflectionClass('SimpleTraitTester'), 'runFromTrait');
+        $this->check(count($annotations) > 0, 'from class reflection and method name');
+
+        $annotations = Annotations::ofMethod(new \ReflectionMethod('SimpleTraitTester', 'runFromTrait'));
+        $this->check(count($annotations) > 0, 'from method reflection');
+
+        $annotations = Annotations::ofMethod(new SimpleTraitTester(), 'runFromTrait');
+        $this->check(count($annotations) > 0, 'from class object and method name');
+
+        $annotations = Annotations::ofMethod('SimpleTraitTester', 'runFromTrait');
+        $this->check(count($annotations) > 0, 'from class name and method name');
+
+        $annotations = Annotations::ofMethod('SimpleTraitTester', 'runFromAnotherTrait');
+        $this->check(count($annotations) > 0, 'from class name and method name for namespaced trait');
+    }
+
+    protected function testHandlesMethodInheritanceWithTraits()
+    {
+        if (version_compare(PHP_VERSION, '5.4.0', '<')) {
+            $this->pass();
+            return;
+        }
+
+        $annotations = Annotations::ofMethod('InheritanceTraitTester', 'baseTraitAndParent');
+        $this->check(count($annotations) === 2, 'baseTraitAndParent inherits parent annotations');
```

Comment:
Ha. That's exactly the case I've mentioned in other comments: traits doesn't inherit, but if classes using them do, then methods sort of gets inherited. So if trait adds method/property to class, that (or it's parent/sub-class) is introspected than annotations are considered to be owned by that class.

Rule of thumb here would be: if you see both echoes printed, then annotations are merged. Here is what `AnnotationManager` should do in regards to this:

1. get annotations of class/trait member (not counting inherited member declarations)
2. if class/trait itself doesn't have that member, then try finding first definition of that member in used traits (matching by name + resolution by `insteadof` and `as`)
3. add up parent class annotations (parent class processing includes both 1. and 2. items)